onnxruntime/core/graph/graph_utils.cc

```diff
@@ -445,6 +530,69 @@ size_t RemoveNodeOutputEdges(Graph& graph, Node& node) {
   return output_edges.size();
 }
 
-}  // namespace graph_utils
+void ReplaceNodeOutput(Graph& graph, Node& node, int output_idx, Node& replacement, int replacement_output_idx) {
+  // get the output edges from node for output_idx
+  std::vector<GraphEdge> output_edges = GetNodeOutputEdges(node, output_idx);
+
+  if (!output_edges.empty()) {
+    const auto& replacement_name = replacement.MutableOutputDefs()[replacement_output_idx]->Name();
+
+    // Remove the output edges of the node first
+    RemoveGraphEdges(graph, output_edges);
+
+    // Create connections between the replacement node and the outgoing nodes
+    for (const auto& output_edge : output_edges) {
+      // Take care of subgraph inputs.
+      if (OutputEdgeProvidesImplicitInput(graph, output_edge)) {
+        Node& mutable_output_edge_node = *graph.GetNode(output_edge.dst_node);
+        UpdateImplicitInputNameInSubgraph(mutable_output_edge_node, output_edge.arg_name, replacement_name);
+      }
+
+      // Add new edge connecting the input with the output nodes directly.
+      // This also updates the destination node's input node args
+      graph.AddEdge(replacement.Index(), output_edge.dst_node, replacement_output_idx, output_edge.dst_arg_index);
+    }
+  }
+}
+
+void ReplaceNodeInput(Node& target, int target_input_idx, NodeArg& replacement) {
+  size_t dst_arg_idx = static_cast<size_t>(target_input_idx);
+  auto num_explicit_inputs = target.InputDefs().size();
+
+  if (dst_arg_idx < num_explicit_inputs) {
+    target.MutableInputDefs()[target_input_idx] = &replacement;
+  } else if (dst_arg_idx < num_explicit_inputs + target.ImplicitInputDefs().size()) {
+    // If we need to update an implicit input.
+    target.MutableImplicitInputDefs()[dst_arg_idx - num_explicit_inputs] = &replacement;
+  } else {
+    // logic error in our code
+    ORT_THROW("Invalid input index for node ", target.Name(), ". Index:", target_input_idx,
+              " ExplicitInputs:", num_explicit_inputs,
+              " ImplicitInputs:", target.ImplicitInputDefs().size());
+  }
+}
+
+void AddNodeInput(Node& target, int target_input_idx, NodeArg& new_input) {
+  auto num_explicit_inputs = target.InputDefs().size();
+  ORT_ENFORCE(target_input_idx == num_explicit_inputs, "Can only add a new input at the end of the current ones.");
 
+  target.MutableInputDefs().push_back(&new_input);
+  assert(target.MutableInputArgsCount().size() > static_cast<size_t>(target_input_idx));  // expect existing entry for all possible inputs
```

Comment:
is this assert different than the ort_enforce above?

Reply:
Yes. The input defs and input args counts are separate structures. I believe there size of the input args count is based on all possible args rather than just the provided ones, but wanted to make that clear with the assert.